### PR TITLE
[AIRFLOW-610] Respect _cmd option in config before defaults

### PR DESCRIPTION
**Pull request by cjgu**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following:
    - https://issues.apache.org/jira/browse/AIRFLOW-610

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
  - This PR is a more or less a rebase of https://github.com/apache/incubator-airflow/pull/1862
  - The command versions of config parameters were overridden by the
default config. E.g `sql_alchemy_conn` got the default value even
when `sql_alchemy_conn_cmd` was specified. 
 -  This PR makes the AirflowConfigParser only look at the default values last by having a separate instance of ConfigParser containing the default values.

### Tests
- [X] My PR adds the following unit tests:
 - `tests/configuration.py:ConfTest.test_command_config`
 - `tests/configuration.py:ConfTest.test_remove_option`

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by cjgu** [comment]
I don't really understand why this test fails on Travis, it works locally with sqlite backend.

```
12) FAIL: test_remove_file_clears_import_error (tests.SchedulerJobTest)
----------------------------------------------------------------------
   Traceback (most recent call last):
    tests/jobs.py line 2889 in test_remove_file_clears_import_error
      self.assertEqual(len(import_errors), 0)
   AssertionError: 1 != 0
``` 
(https://travis-ci.org/apache/incubator-airflow/jobs/340156757#L4496)

File: airflow/configuration.py [hunk 1]
```diff
@@ -197,15 +186,20 @@ def get(self, section, key, **kwargs):
             return option
 
         # ...then the config file
-        if self.has_option(section, key):
+        if super(AirflowConfigParser, self).has_option(section, key):
```
Comment (ashb) [inline airflow/configuration.py:189]: This seems an odd thing to do -- can we not just use `self.has_option(section, key)` here? If not expand in the comment as to why please.

File: airflow/configuration.py [hunk 2]
```diff
@@ -235,9 +229,28 @@ def getfloat(self, section, key):
...
+            return True
+        except AirflowConfigException:
+            return False
+
+    def remove_option(self, section, option):
+        """
+        Remove an option if it exists in config from a file or
+        default config. If both of config have the same option, this remove
+        the option in config loaded from a file at first and remove it from
+        default config at next call.
```
Comment (ashb) [inline airflow/configuration.py]: I think a clearer api here would be to have this method be `configuration.remove_option('core', 'something', remove_default=True)` - rather than having to call it twice.

File: airflow/configuration.py [hunk 3]
```diff
@@ -418,55 +430,30 @@ def parameterized_config(template):
...
-as_dict.__doc__ = conf.as_dict.__doc__
-
+    conf.load_test_config()
 
-def set(section, option, value):  # noqa
-    return conf.set(section, option, value)
+# Historical convenience functions to access config entries
+
+load_test_config = conf.load_test_config
+get = conf.get
```
Comment (ashb) [inline airflow/configuration.py:464]: Nice pattern.

File: tests/core.py [hunk 4]
```diff
@@ -690,6 +690,9 @@ def test_config_throw_error_when_original_and_fallback_is_absent(self):
         self.assertFalse(configuration.has_option("core", "FERNET_KEY_CMD"))
 
         FERNET_KEY = configuration.get("core", "FERNET_KEY")
+        # remove overridden config
+        configuration.remove_option("core", "FERNET_KEY")
+        # remove default config
```
Comment (ashb) [inline tests/core.py]: See my comment about API on remove_option.

File: airflow/configuration.py [hunk 5]
```diff
@@ -235,9 +229,28 @@ def getfloat(self, section, key):
         return float(self.get(section, key))
 
     def read(self, filenames):
-        ConfigParser.read(self, filenames)
+        super(AirflowConfigParser, self).read(filenames)
         self._validate()
 
+    def has_option(self, section, option):
+        try:
+            self.get(section, option)
```
Comment (ashb) [inline airflow/configuration.py]: This is going to run any commands. How much would would be to check for the `_cmd` style instead -- i.e. to not call `.get()` here.

If a lot then add a comment saying why we are doing a get check and catching the error

File: tests/configuration.py [hunk 6]
```diff
@@ -55,6 +57,27 @@ def test_conf_as_dict(self):
...
+
+        test_conf = AirflowConfigParser(
+            default_config=parameterized_config(TEST_CONFIG_DEFAULT))
+        test_conf.read_string(TEST_CONFIG)
+        test_conf.as_command_stdout = test_conf.as_command_stdout | {
+            ('test', 'key2')
+        }
+        self.assertEqual('hello', test_conf.get('test', 'key1'))
+        self.assertEqual('cmd_result', test_conf.get('test', 'key2'))
+        self.assertEqual('airflow', test_conf.get('test', 'key3'))
```
Comment (ashb) [inline tests/configuration.py]: Add a test that covers `has_option` when using cmd style. Possibly also when the non_cmd version is not in the default config.

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by ashb** [comment]
Oh, can you rename this PR to something like "Correctly respect _cmd option in config files above default values" (if that isn't too long)

**Comment by cjgu** [comment]
Thank you @ashb, I will revise the PR!

**Comment by cjgu** [comment]
Thank you @ashb, I will revise the PR!

File: tests/configuration.py (hunk 6 above)
Comment (cjgu) [reply tests/configuration.py]: Added!

File: airflow/configuration.py (hunk 2 above)
Comment (cjgu) [reply airflow/configuration.py]: Changed the api to what you suggested.

File: airflow/configuration.py (hunk 1 above)
Comment (cjgu) [reply airflow/configuration.py:189]: Yeah, so as `self.has_option` uses `self.get` we would get infinite recursion here. Also self.has_option also would get defaults options already here. I've added comments to the code to address that. 

We could reimplement has_option but it would not be usable here anyway as it would look not only at the config file but env variables and the default config. It would then have to have some flag to avoid that.

**Comment by cjgu** [comment]
@ashb I addressed most comments now and manage to fix an actual bug with my implementation that the previous failing test catched. Please take a second look!

File: tests/configuration.py [hunk 7]
```diff
@@ -55,6 +57,57 @@ def test_conf_as_dict(self):
...
+        test_conf = AirflowConfigParser(
+            default_config=parameterized_config(TEST_CONFIG_DEFAULT))
+        test_conf.read_string(TEST_CONFIG)
+
+        self.assertEqual('hello', test_conf.get('test', 'key1'))
+        test_conf.remove_option('test', 'key1', remove_default=False)
+        self.assertEqual('awesome', test_conf.get('test', 'key1'))
+
+        test_conf.remove_option('test', 'key2')
+        self.assertEqual(False, test_conf.has_option('test', 'key2'))
```
Comment (ashb) [inline tests/configuration.py]: `self.assertFalse()` please.

File: tests/configuration.py [hunk 8]
```diff
@@ -55,6 +57,57 @@ def test_conf_as_dict(self):
...
+        self.assertEqual('hello', test_conf.get('test', 'key1'))
+        self.assertEqual('cmd_result', test_conf.get('test', 'key2'))
+        self.assertEqual('airflow', test_conf.get('test', 'key3'))
+        self.assertEqual('key4_result', test_conf.get('test', 'key4'))
+
+        self.assertEqual(True, test_conf.has_option('test', 'key1'))
+        self.assertEqual(True, test_conf.has_option('test', 'key2'))
+        self.assertEqual(True, test_conf.has_option('test', 'key3'))
+        self.assertEqual(True, test_conf.has_option('test', 'key4'))
+        self.assertEqual(False, test_conf.has_option('test', 'key5'))
```
Comment (ashb) [inline tests/configuration.py]: `assertTrue`/`assertFalse` here please.

File: airflow/configuration.py [hunk 9]
```diff
@@ -418,55 +436,30 @@ def parameterized_config(template):
...
+set = conf.set # noqa
+
+for func in [load_test_config, get, getboolean, getfloat, getint, has_option,
+             remove_option, as_dict, set]:
+    _deprecated(
+        func,
+        "Accessing configuration method '{f.__name__}' directly from "
+        "the configuration module is deprecated. Please access the "
+        "configuration from the 'configuration.conf' object via "
+        "'conf.{f.__name__}'".format(f=func))
```
Comment (ashb) [inline airflow/configuration.py:481]: Is this a new deprecation? (I think it is), and if so do we trigger this anywhere from within the Airflow code-base? If so probably best to make that a separate change.

If we don't currently trigger this warning from this repo then this change can stay.

Comment (ashb) [reply airflow/configuration.py:481]: Looks like we would trigger this in a few places: `with timeout(configuration.getint('core', "DAGBAG_IMPORT_TIMEOUT")):` in airflow/models.py etc.

(I am -0.1 to this change, but don't mind it so long as we don't trigger our own deprecation warnings.)

**Review by ashb — CHANGES_REQUESTED** [review]
A few more minor fixes/changes. Sorry for not catching them all first time.

File: airflow/configuration.py (hunk 9 above)
Comment (cjgu) [reply airflow/configuration.py:481]: So it was suggested in the original PR that I based it on, I'm happy to remove the deprecation stuff if you want. See https://github.com/apache/incubator-airflow/pull/1862#issuecomment-260429818

File: tests/configuration.py (hunk 7 above)
Comment (cjgu) [reply tests/configuration.py]: Fixed.

File: tests/configuration.py (hunk 8 above)
Comment (cjgu) [reply tests/configuration.py]: Fixed.

File: airflow/configuration.py (hunk 9 above)
Comment (ashb) [reply airflow/configuration.py:481]: Ah cool, lets keep it in then, but try to fix up any references in the rest of the code to the deprecated versions.

Comment (cjgu) [reply airflow/configuration.py:481]: Ok, I made a separate commit for that so it is easier to review. I can squash them when reviewed.

**Review by ashb — APPROVED** [review]
Good work. Once tests are green please squash.

**Comment by cjgu** [comment]
Thanks! Squashed now.

**Comment by cjgu** [comment]
What's the process for getting this merged? @ashb

**Comment by ashb** [comment]
@bolkedebruin @jgao54 @Fokko Can we get a merge on this please?

**Comment by cjgu** [comment]
I found an edge case which my patch does not handle that probably should be fixed before merging, if a section is missing in the config but exists in default it gives an error, so lets wait for that. Will fix that during the weekend.

**Comment by ashb** [comment]
@cjgu Did you get a chance to fix the edge case you found?

**Comment by cjgu** [comment]
Sorry, life got in the way, but I haven't forgotten about it. I have a test case to reproduce the issue, now just need to actually fix it. Might have some time this weekend.

**Comment by ashb** [comment]
Good - Life is important too! Just going through recent PRs that I was involved in now I can merge them directly. No rush.

**Comment by cjgu** [comment]
@ashb Please take a look at the commit I added, will squash it when approved!

**Comment by cjgu** [comment]
Great, squashed it now, so should be ready for merge.

**Comment by cjgu** [comment]
Seems there's some flaky tests, rerunning the tests.

**Review by Fokko — APPROVED** [review]
I like this, makes the behaviour of the config much more predictable.

**Comment by cjgu** [comment]
I rebased this onto latest master and fixed the conflicts. It would be nice to get this merged soon as it touches quite a lot of files and easily attracts conflicts. @ashb @Fokko

**Comment by cjgu** [comment]
Sorry for bugging you, but I really would like to get this merged soon as it's been ready for a few weeks now. @ashb @Fokko

**Comment by cjgu** [comment]
Ugh looked like I rebased at a bad point, will redo it.

**Comment by cjgu** [comment]
Looks like the failing tests fail on other branches as well. So this should be ready to be merged.

**Comment by Fokko** [comment]
@cjgu Please rebase on master once more. Master is green now, since so many files are touched, we need to make sure that it is ok.

**Comment by cjgu** [comment]
@Fokko Yeah makes sense. It is now rebased, and the tests passes.

**Comment by Fokko** [comment]
One more question, why adding the `.conf` everywhere. The without it works also, and is more concise.

**Comment by cjgu** [comment]
@Fokko That's due to feedback from the previous PR, https://github.com/apache/incubator-airflow/pull/1862. (more specifically https://github.com/apache/incubator-airflow/pull/1862#issuecomment-260429818)

**Comment by cjgu** [comment]
Any more feedback @Fokko or could we merge this?

**Review by Fokko — APPROVED** [review]
LGTM, Thanks @cjgu

**Comment by cjgu** [comment]
Thank you @Fokko!
